_posts/2020-03-15-method-s-arguments-pt-1.markdown

```diff
@@ -0,0 +1,373 @@
+
+# Required positional arguments
+
+These positional arguments are required when calling a method, it's not that it's required to use them when defining the method (you can have no arguments, only optionals, etc). If the method defines required parameters you have to provide an argument for each position.
+
+```ruby
+def foo(arg1)
+  puts "arg1 is: #{arg1.inspect}"
+end
+
+foo(1)
+# => arg1 is: 1
+
+foo
+# ArgumentError (wrong number of arguments (given 0, expected 1))
+```
+
+We can have as many as we want:
+
+```ruby
+def foo(arg1, arg2, arg3, arg4)
+# and more and more...
+```
+
+> Note that methods with too many arguments are a sign of bad design since the method is probably doing too many things!
+
+# Optional positional arguments
+
+Sometimes we may want to allow the user to provide more input but not require them to do so. For that, we can use optional arguments for which we define a default value in case the user doesn't specify one. This is particularly useful when we want to have some default behavior for our method but give the user the option to modify that. If we just hardcoded the default value inside the method we wouldn't have this flexibility.
+
+```ruby
+def foo(arg1 = 'default value')
+  puts "arg1 is: #{arg1.inspect}"
+end
+
+foo(1)
+# => arg1 is: 1
+
+foo
+# => arg1 is: default value
+```
+
+We can have as many as we want, just like the required parameters.
+
+# Combining required and optional positional arguments
+
+Now we can combine both required and positional arguments in one method definition:
+
+```ruby
+def foo(arg1, arg2 = 'default value')
+  puts "arg1 es: #{arg1.inspect}"
+  puts "arg2 es: #{arg2.inspect}"
+end
+
+foo(1)
+# => arg1 is: 1
+#    arg2 is: default value
+
+foo(1,2)
+# => arg1 is: 1
+#    arg2 is: 2
+```
+
+Usually, when we mix both types of arguments, we put the required arguments first and then the optional ones. But Ruby allows us to do weird things like:
+
+```ruby
+def foo(arg1, arg2 = 'default value', arg3)
+  puts "arg1 is: #{arg1.inspect}"
+  puts "arg2 is: #{arg2.inspect}"
+  puts "arg3 is: #{arg3.inspect}"
+end
+```
+
+But this can create confusing situations:
+
+```ruby
+foo(1, 2, 3)
+# => arg1 is: 1
+#    arg2 is: 2
+#    arg3 is: 3
+
+foo(1,2)
+# => arg1 is: 1
+#    arg2 is: default value
+#    arg3 is: 2
+```
+
+Notice that, on the second call, the second argument ends up at **arg3**. My first thought was that ruby first assigned the required arguments and then the optionals, but that's not the case. Check out this even more confusing example:
+
+```ruby
+def foo(arg1, arg2 = 'default value X', arg3 = 'default value Y', arg4)
+  puts "arg1 is: #{arg1.inspect}"
+  puts "arg2 is: #{arg2.inspect}"
+  puts "arg3 is: #{arg3.inspect}"
+  puts "arg4 is: #{arg4.inspect}"
+end
+
+foo(1, 2, 3, 4)
+arg1 is: 1
+arg2 is: 2
+arg3 is: 3
+arg4 is: 4
```

Comment:
@arielj It might display in a weird way because your code snippet is using `ruby` -- is there any way to improve syntax highlighting here? 

![Screen Shot 2020-04-08 at 10 23 54 AM](https://user-images.githubusercontent.com/17584/78795503-18aeb180-7983-11ea-91ef-2702237da236.png)

Reply:
yes, I forgot the # at te beginning like I used on the other snippets